Source/Core/Core/Lua/LuaFunctions/LuaRegisters.cpp

```diff
@@ -0,0 +1,467 @@
+#include "LuaRegisters.h"
+#include "../LuaVersionResolver.h"
+#include "../LuaHelperClasses/NumberType.h"
+#include "common/CommonTypes.h"
+#include "Core/PowerPC/PowerPC.h"
+#include <algorithm>
+#include <string>
+
+namespace Lua
+{
+namespace LuaRegisters
+{
+class luaRegister
+{
+public:
+  inline luaRegister() {}
+};
+
+luaRegister* luaRegisterPointer = NULL;
+
+luaRegister* GetLuaRegisterInstance()
+{
+  if (luaRegisterPointer == NULL)
+    luaRegisterPointer = new luaRegister();
+  return luaRegisterPointer;
+}
+
+// Currently supported registers are:
+// r0 - r31, f0 - f31, PC, and LR (the register which stores the return address to jump to when a
+// function call ends)
+void InitLuaRegistersFunctions(lua_State* luaState, const std::string& luaApiVersion)
+{
+  luaRegister** luaRegisterPtrPtr = (luaRegister**)lua_newuserdata(luaState, sizeof(luaRegister*));
+  *luaRegisterPtrPtr = GetLuaRegisterInstance();
+  luaL_newmetatable(luaState, "LuaRegistersTable");
+  lua_pushvalue(luaState, -1);
+  lua_setfield(luaState, -2, "__index");
+
+  std::array luaRegistersFunctionsWithVersionsAttached = {
+    luaL_Reg_With_Version({"getRegister", "1.0", getRegister}),
+    luaL_Reg_With_Version({"getRegisterAsUnsignedByteArray", "1.0", getRegisterAsUnsignedByteArray}),
+    luaL_Reg_With_Version({"getRegisterAsSignedByteArray", "1.0", getRegisterAsSignedByteArray}),
+    luaL_Reg_With_Version({"setRegister", "1.0", setRegister}),
+    luaL_Reg_With_Version({"setRegisterFromByteArray", "1.0", setRegisterFromByteArray})
+  };
+
+  std::unordered_map<std::string, std::string> deprecatedFunctionsMap = std::unordered_map<std::string, std::string>();
+
+  addLatestFunctionsForVersion(luaRegistersFunctionsWithVersionsAttached, luaApiVersion, deprecatedFunctionsMap, luaState);
+  lua_setglobal(luaState, "registers");
+}
+
+class RegisterObject
+{
+public:
+
+  enum REGISTER_TYPE
```

Comment:
We don't use all-caps for type names. This can also be an enum class

```suggestion
  enum class RegisterType
```

similarly, the members should also be `CamelCase`